packages/react/src/components/AccountSettings/ChangePassword/ChangePassword.tsx

```diff
@@ -1,53 +1,163 @@
+    return validators ?? getDefaultPasswordValidators();
+  }, [validators]);
+
+  /**
+   * Note that formValues and other states are passed in as props so that
+   * it does not depend on whether or not those states have been updated yet
+   */
+  const validateNewPassword = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword } = formValues;
+
+      return runFieldValidators(
+        newPassword,
+        passwordValidators,
+        eventType,
+        blurredFields.newPassword
+      );
+    },
+    [passwordValidators]
+  );
+
+  const validateConfirmPassword = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ): string[] => {
+      const { newPassword, confirmPassword } = formValues;
+
+      const confirmPasswordValidator = getConfirmPasswordValidator(newPassword);
+
+      return runFieldValidators(
+        confirmPassword,
+        [confirmPasswordValidator],
+        eventType,
+        blurredFields.confirmPassword
+      );
+    },
+    []
+  );
+
+  const runValidation = React.useCallback(
+    (
+      formValues: FormValues,
+      blurredFields: BlurredFields,
+      eventType: InputEventType
+    ) => {
+      const newPasswordErrors = validateNewPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+      const confirmPasswordErrors = validateConfirmPassword(
+        formValues,
+        blurredFields,
+        eventType
+      );
+
+      const newValidationError = {
+        newPassword: newPasswordErrors,
+        confirmPassword: confirmPasswordErrors,
+      };
+
+      // only re-render if errors have changed
+      if (!isEqual(validationError, newValidationError)) {
+        setValidationError(newValidationError);
+      }
+    },
+    [validateConfirmPassword, validateNewPassword, validationError]
+  );
 
   /** Translations */
   // TODO: add AccountSettingsTextUtil to collect these strings
   const currentPasswordLabel = translate('Current Password');
   const newPasswordLabel = translate('New Password');
+  const confirmPasswordLabel = translate('Confirm Password');
   const updatePasswordText = translate('Update password');
 
   /** Event Handlers */
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     event.preventDefault();
 
     const { name, value } = event.target;
-    setFormValues({ ...formValues, [name]: value });
+
+    const newFormValues = { ...formValues, [name]: value };
+    runValidation(newFormValues, blurredFields, 'change');
+
+    setFormValues(newFormValues);
+  };
+
+  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+    event.preventDefault();
+
+    const { name } = event.target;
+    const newBlurredFields = { ...blurredFields, [name]: true };
+    runValidation(formValues, newBlurredFields, 'blur');
```

Comment:
Wonder if this should be ran as an effect rather than called here, thoughts?

Reply:
IMO no, because this also depends on what the inputEvent was beyond React states `formValues` and `newBlurredFields`. We'll have to keep track of last `inputEvent` type if we want to convert this to an effect. 

I personally think having this as util makes it easier to see when it is triggered from which event.

Reply:
Leaving this as-is, and open a separate ticket to refactor this later, as discussed offline.